File: packages/swift/Sources/PylonSync/MutationQueue.swift
import Foundation
import PylonClient

public struct PendingMutation: Sendable, Codable, Hashable {
    public var id: String
    public var change: ClientChange
    public var status: Status
    public var error: String?
    /// Pre-mutation snapshot of the row, captured at queue time for
    /// update/delete. Used to restore the row if the push is permanently
    /// rejected (see `SyncEngine.failPushedMutation`). `nil` for inserts and
    /// for updates whose target was itself an un-acked insert.
    public var prevRow: Row?

    public enum Status: String, Sendable, Codable, Hashable {
        case pending
        case applied
        case failed
    }

    public init(
        id: String,
        change: ClientChange,
        status: Status = .pending,
        error: String? = nil,
        prevRow: Row? = nil
    ) {
        self.id = id
        self.change = change
        self.status = status
        self.error = error
        self.prevRow = prevRow
    }
}

/// Persistence backend for the mutation queue. Implement against SQLite,
/// the filesystem, or any KV store. The default sync engine swaps in
/// `SQLiteMutationPersistence` from this module.
public protocol MutationQueuePersistence: Sendable {
    func saveAll(_ mutations: [PendingMutation]) async throws
    func loadAll() async throws -> [PendingMutation]
}

/// Offline-safe write queue. Mutations are minted with a stable `op_id`
/// that doubles as the server-side idempotency key — replays on retry are
/// short-circuited server-side.
///
/// Mirrors `MutationQueue` from `packages/sync/src/index.ts`. Failed
/// mutations are kept (not dropped) so the UI can surface them to the
/// user. Applied mutations are pruned by `clear()`.
public actor MutationQueue {
    private var queue: [PendingMutation] = []
    private var persistence: MutationQueuePersistence?

    public init(persistence: MutationQueuePersistence? = nil) {
        self.persistence = persistence
    }

    public func attachPersistence(_ p: MutationQueuePersistence) {
        self.persistence = p
    }

    /// Load any persisted mutations from the backend. Call once at startup.
    public func hydrate() async {
        guard let persistence else { return }
        do {
            let loaded = try await persistence.loadAll()
            let existing = Set(queue.map(\.id))
            var mergedAny = false
            for m in loaded where !existing.contains(m.id) {
                queue.append(m)
                mergedAny = true
            }
            if mergedAny { await flush() }
        } catch {
            // Broken storage shouldn't take down the app — degrade to
            // memory-only mode silently. App can inspect logs to diagnose.
        }
    }

    /// Append a mutation. Returns the `op_id` for caller bookkeeping.
    /// `prevRow` is the pre-mutation row snapshot (update/delete) used to
    /// roll back on a permanent push rejection.
    @discardableResult
    public func add(_ change: ClientChange, prevRow: Row? = nil) async -> String {
        let id = "mut_\(Int(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString.prefix(8))"
        var changeWithOp = change
        changeWithOp.op_id = id
        queue.append(PendingMutation(id: id, change: changeWithOp, prevRow: prevRow))
        await flush()
        return id
    }

    public func pending() -> [PendingMutation] {
        queue.filter { $0.status == .pending }
    }

    public func all() -> [PendingMutation] { queue }

    /// `"{entity}/{row_id}"` keys for every pending OR failed mutation.
    /// Reconcile must NOT sweep or overwrite these rows — an offline insert
    /// that hasn't pushed yet would otherwise look like a phantom local row
    /// and get tombstoned before it ships. Mirrors TS `pendingRowKeys()`.
    public func pendingRowKeys() -> Set<String> {
        var keys = Set<String>()
        for m in queue where m.status == .pending || m.status == .failed {
            keys.insert("\(m.change.entity)/\(m.change.row_id)")
        }
        return keys
    }

    public func markApplied(_ id: String) async {
        if let idx = queue.firstIndex(where: { $0.id == id }) {
            queue[idx].status = .applied
        }
        await flush()
    }

    public func markFailed(_ id: String, error: String) async {
        if let idx = queue.firstIndex(where: { $0.id == id }) {
            queue[idx].status = .failed
            queue[idx].error = error
        }
        await flush()
    }

    /// Drop applied mutations. Failed ones are kept so the UI can ack/retry.
    public func clear() async {
        queue.removeAll { $0.status == .applied }
        await flush()
    }

    /// Drop EVERY queued mutation (pending + failed) and clear the persisted
    /// copy. Used on an identity flip: the outgoing identity's un-pushed
    /// writes must NOT be replayed under the incoming identity's token — the
    /// "cross-identity write leak" the TS client's `wipeMutations` prevents.
    public func wipeAll() async {
        queue.removeAll()
        await flush()
    }

    public func remove(_ id: String) async {
        queue.removeAll { $0.id == id }
        await flush()
    }

    private func flush() async {
        guard let persistence else { return }
        let snapshot = queue
        do {
            try await persistence.saveAll(snapshot)
        } catch {
            // Log + drop — see comment on hydrate(). The next mutation will
            // re-attempt the write.
        }
    }
}

File: packages/sync/src/mutation-queue.ts
// ---------------------------------------------------------------------------
// Offline-safe write queue.
//
// Memory-only by default. When a persistence backend is attached
// (`attachPersistence`), every state transition writes through to
// disk and `hydrate()` restores pending/failed mutations on startup.
//
// The mutation id is reused as the server-side `op_id` for idempotent
// replay — a retry carrying the same id short-circuits on the server
// instead of re-applying.
// ---------------------------------------------------------------------------

import type { ClientChange } from "./types";

export interface PendingMutation {
  id: string;
  change: ClientChange;
  status: "pending" | "applied" | "failed";
  error?: string;
}

/**
 * Optional persistence backend. The default IndexedDB persistence
 * layer provides `savePending`/`loadPending`. Callers can supply a
 * custom backend for tests or alternative storage.
 */
export interface MutationQueuePersistence {
  saveAll(mutations: PendingMutation[]): Promise<void>;
  loadAll(): Promise<PendingMutation[]>;
}

export class MutationQueue {
  private queue: PendingMutation[] = [];
  private persistence?: MutationQueuePersistence;

  constructor(persistence?: MutationQueuePersistence) {
    this.persistence = persistence;
  }

  /**
   * Attach a persistence backend after construction. The SyncEngine
   * swaps in IndexedDB-backed persistence once the DB has opened.
   * Public so it doesn't need a `// @ts-expect-error` to reach in
   * from the same package.
   */
  attachPersistence(persistence: MutationQueuePersistence): void {
    this.persistence = persistence;
  }

  /** Load persisted queue state. Call once at startup. */
  async hydrate(): Promise<void> {
    if (!this.persistence) return;
    try {
      const loaded = await this.persistence.loadAll();
      // Merge in-memory with on-disk. An `add()` that ran while
      // hydrate was awaiting `loadAll()` will already have flushed a
      // snapshot that didn't include the loaded rows — re-flush
      // after merge so disk matches memory again.
      const existingIds = new Set(this.queue.map((m) => m.id));
      let mergedAny = false;
      for (const m of loaded) {
        if (!existingIds.has(m.id)) {
          this.queue.push(m);
          mergedAny = true;
        }
      }
      if (mergedAny) this.flush();
    } catch (err) {
      // Broken storage shouldn't prevent the app from running — warn
      // and degrade to memory-only mode.
      console.warn("[sync] mutation-queue hydrate failed:", err);
    }
  }

  /** Add a pending mutation. Returns the op_id used for server
   *  idempotency. */
  add(change: ClientChange): string {
    const id = `mut_${Date.now()}_${Math.random().toString(36).slice(2)}`;
    const changeWithOp: ClientChange = { ...change, op_id: id };
    this.queue.push({ id, change: changeWithOp, status: "pending" });
    this.flush();
    return id;
  }

  pending(): PendingMutation[] {
    return this.queue.filter((m) => m.status === "pending");
  }

  /**
   * Set of `${entity}/${row_id}` keys for every mutation currently
   * in Pending or Failed state. Used by reconcile() to skip rows
   * whose canonical state on the server hasn't caught up with the
   * local optimistic ghost yet — otherwise reconcile would tombstone
   * the row (it's not yet on the server) and the still-pending push
   * would later re-apply against the tombstone, fighting the local
   * replica.
   *
   * Failed mutations are included too: a user-visible failure is
   * recoverable, and sweeping the row would discard the local edit
   * the user is still trying to push.
   */
  pendingRowKeys(): Set<string> {
    const out = new Set<string>();
    for (const m of this.queue) {
      if (m.status === "pending" || m.status === "failed") {
        out.add(`${m.change.entity}/${m.change.row_id}`);
      }
    }
    return out;
  }

  markApplied(id: string): void {
    const m = this.queue.find((m) => m.id === id);
    if (m) m.status = "applied";
    this.flush();
  }

  markFailed(id: string, error: string): void {
    const m = this.queue.find((m) => m.id === id);
    if (m) {
      m.status = "failed";
      m.error = error;
    }
    this.flush();
  }

  /**
   * Prune applied mutations. Failed mutations are KEPT so the UI can
   * surface them to the user and so retries are possible.
   */
  clear(): void {
    this.queue = this.queue.filter(
      (m) => m.status === "pending" || m.status === "failed",
    );
    this.flush();
  }

  /** Remove a specific mutation by id. Used by the UI after user
   *  ack of failures. */
  remove(id: string): void {
    this.queue = this.queue.filter((m) => m.id !== id);
    this.flush();
  }

  /** Fire-and-forget persistence write. */
  private flush(): void {
    if (!this.persistence) return;
    const snapshot = this.queue.slice();
    this.persistence.saveAll(snapshot).catch((err) => {
      console.warn("[sync] mutation-queue persist failed:", err);
    });
  }
}
